### PR TITLE
Rework custom operator docs

### PR DESCRIPTION
**Pull request by klecki**
Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

#### Why we need this PR?
- Docs rework for custom ops

#### What happened in this PR?
 - What solution was applied:
	Custom Op C++ was reworked for the SetupImpl/RunImpl API and sane workspace.
    torch_python_function was exposed as fn API and enabled in the documentation.
    CPU and GPU python tutorials reworked with fn API.     
- Affected modules and functionalities:
     Docs for custom ops.
 - Key points relevant for the review:
     Probably spelling
 - Validation and testing:
     CI, will look at renders.
 - Documentation (including examples):
     Yes.

**JIRA TASK**: *[Use DALI-1758]*

### CODE REVIEW

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

File: docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb [hunk 1]
```diff
@@ -4,30 +4,33 @@
...
-    "- You want to optimize your pipeline by providing a manually fused operation in C++\n",
+    "- DALI does not support the operation that you want to perform and it cannot be expressed by a composition of other operators.\n",
+    "- You want to write an operator that depends on a third-party library.\n",
+    "- You want to optimize your pipeline by providing a manually fused operation in C++.\n",
     "\n",
-    "In this tutorial we will walk you through the process of writing, compiling and loading a plugin with a DALI custom operator. For demonstration purposes we will provide a CPU and a GPU implementation for the 'CustomDummy' operator. The implementation will just copy the input data to the output without any modifications.\n",
+    "In this tutorial, we will walk you through the process of writing, compiling, and loading a plugin with a DALI custom operator. For demonstration purposes we will provide a CPU and a GPU implementation for the CustomDummy operator. The implementation only copies the input data to the output without any modifications.\n",
     "\n",
-    "PREREQUISITES:\n",
+    "### Prerequisites\n",
```
Comment (klecki) [inline docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb:17]: This level of nesting broke the navigation, it seems that `##` needs to be introduced first.

File: docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb [hunk 2]
```diff
@@ -109,25 +117,31 @@
...
       "\r\n",
       "}  // namespace other_ns\r\n",
       "\r\n",
       "DALI_REGISTER_OPERATOR(CustomDummy, ::other_ns::Dummy<::dali::CPUBackend>, ::dali::CPU);\r\n",
       "\r\n",
-      "DALI_SCHEMA(CustomDummy)\r\n",
-      "  .DocStr(\"Make a copy of the input tensor\")\r\n",
-      "  .NumInput(1)\r\n",
-      "  .NumOutput(1);\r\n"
+      "DALI_SCHEMA(CustomDummy).DocStr(\"Make a copy of the input tensor\").NumInput(1).NumOutput(1);\r\n"
```
Comment (banasraf) [inline docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb:144]:
```c++
DALI_SCHEMA(CustomDummy)
  .DocStr("Make a copy of the input tensor")
  .NumInput(1)
  .NumOutput(1);
```
seems to be more readable for me.

File: docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb [hunk 3]
```diff
@@ -141,7 +155,12 @@
    "source": [
     "## GPU operator implementation\n",
     "\n",
-    "Similarly, we provide a GPU implementation in a CUDA implementation file and register it with DALI_REGISTER_OPERATOR."
+    "1. Provide a GPU implementation in a CUDA implementation file by overriding the RunImpl method for DeviceWorkspace.\n",
+    "2. Register the GPU version of operator with DALI_REGISTER_OPERATOR macro.\n",
```
Comment (banasraf) [inline docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb:159]:
```suggestion
    "2. Register the GPU version of the operator with DALI_REGISTER_OPERATOR macro.\n",
```
?

File: docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb [hunk 4]
```diff
@@ -91,9 +94,14 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "## CPU operator implementation\n",
+    "## CPU Operator Implementation\n",
+    "\n",
+    "1. Provide the CPU implementation in a C++ implementation file by overriding the RunImpl method for HostWorkspace.\n",
+    "2. Register the schema for the custom operator with DALI_SCHEMA macro and register the CPU version of the operator with DALI_REGISTER_OPERATOR.\n",
+    "\n",
+    "In RunImpl we obtain access to the entire batch that is processed. We get the reference to the CPU thread pool from the workspace `ws` and create tasks that will copy samples from input to output in parallel. The tasks will be orderd by the thread pool from the longest to the shortest, based on the tensor size, to best utilize the worker threads.\n",
```
Comment (jantonguirao) [inline docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb:102]:
```suggestion
    "In RunImpl we obtain access to the entire batch that is processed. We get the reference to the CPU thread pool from the workspace `ws` and create tasks that will copy samples from input to output in parallel. The tasks will be ordered by the thread pool from the longest to the shortest, based on the tensor size, to best utilize the worker threads.\n",
```

File: docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb [hunk 5]
```diff
@@ -141,7 +155,12 @@
    "source": [
     "## GPU operator implementation\n",
     "\n",
-    "Similarly, we provide a GPU implementation in a CUDA implementation file and register it with DALI_REGISTER_OPERATOR."
+    "1. Provide a GPU implementation in a CUDA implementation file by overriding the RunImpl method for DeviceWorkspace.\n",
+    "2. Register the GPU version of operator with DALI_REGISTER_OPERATOR macro.\n",
+    "\n",
+    "As it was the case for the CPU implementation, we obtain the entire batch in the RunImpl function. The outputs are already allocated based on the return value of SetupImpl function that was provided earlier.\n",
+    "\n",
+    "It is important that we issue the GPU operations on the stream provided through workspace. Here we copy the batch using cudaMemcpyAsync."
```
Comment (jantonguirao) [inline docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb:163]:
```suggestion
    "It is important that we issue the GPU operations on the stream provided by the workspace. Here we copy the batch using cudaMemcpyAsync."
```

File: docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb [hunk 6]
```diff
@@ -4,30 +4,33 @@
    "cell_type": "markdown",
```
Comment (jantonguirao) [inline docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb:4]: <p>Prerequisites is not rendered?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA/DALI/pull/2568/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

File: docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb (hunk 4 above)
Comment (klecki) [reply docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb:102]: done

File: docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb (hunk 2 above)
Comment (klecki) [reply docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb:144]: Done

File: docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb (hunk 3 above)
Comment (klecki) [reply docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb:159]: done

File: docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb (hunk 5 above)
Comment (klecki) [reply docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb:163]: Done

File: docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb [hunk 7]
```diff
@@ -317,7 +336,7 @@
...
    "metadata": {},
    "source": [
-    "Now we are ready to compile our plugin containing the custom operator 'CustomDummy'"
+    "3. We are now ready to compile the plugin that contains the CustomDummy custom operator."
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb:339]: Done everywhere.

File: docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb [hunk 8]
```diff
@@ -402,7 +421,7 @@
    "source": [
     "## Importing the plugin\n",
     "\n",
-    "First we can see that there is no such plugin named 'CustomDummy'"
+    "1. We can see that there is no such operator called CustomDummy."
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/custom_operations/custom_operator/create_a_custom_operator.ipynb:424]: Done

File: docs/examples/custom_operations/custom_operator/customdummy/dummy.cc [hunk 9]
```diff
@@ -2,22 +2,28 @@
...
 
 }  // namespace other_ns
 
 DALI_REGISTER_OPERATOR(CustomDummy, ::other_ns::Dummy<::dali::CPUBackend>, ::dali::CPU);
 
-DALI_SCHEMA(CustomDummy)
-  .DocStr("Make a copy of the input tensor")
-  .NumInput(1)
-  .NumOutput(1);
+DALI_SCHEMA(CustomDummy).DocStr("Make a copy of the input tensor").NumInput(1).NumOutput(1);
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/custom_operations/custom_operator/customdummy/dummy.cc:29]: Done.

**Comment by klecki** [comment]
!build

**Review by banasraf — APPROVED** [review]

**Comment by klecki** [comment]
!build

**Review by jantonguirao — APPROVED** [review]

**Comment by klecki** [comment]
!build

**Pull request merged**